ValveResourceFormat/KeyValues/KeyValues3.cs

```diff
@@ -0,0 +1,526 @@
+                }
+                return;
+            }
+
+            parser.currentString.Append(c);
+        }
+
+        //Read an array
+        private static void ReadValueArray(char c, Parser parser)
+        {
+            //This shouldn't happen
+            if (!Char.IsWhiteSpace(c) && c != ',')
+            {
+                throw new Exception("Error parsing array.");
+            }
+
+            //Just jump to seek_value state
+            parser.stateStack.Push(State.SEEK_VALUE);
+        }
+
+        //Read a flagged value
+        private static void ReadValueFlagged(char c, Parser parser)
+        {
+            //End at whitespace
+            if (Char.IsWhiteSpace(c))
+            {
+                parser.stateStack.Pop();
+                parser.objStack.Peek().AddProperty(parser.currentName, new KVValue(KVType.FLAGGED_STRING, parser.currentString.ToString()));
+                return;
+            }
+
+            parser.currentString.Append(c);
+        }
+
+        //Read comments
+        private static void ReadComment(char c, Parser parser)
+        {
+            //Check for multiline comments
+            if (parser.currentString.Length == 1 && c == '*')
+            {
+                parser.stateStack.Pop();
+                parser.stateStack.Push(State.COMMENT_BLOCK);
+            }
+
+            //Check for the end of a comment
+            if (c == '\n')
+            {
+                parser.stateStack.Pop();
+                return;
+            }
+
+            if (c != '\r')
+            {
+                parser.currentString.Append(c);
+            }
+        }
+
+        //Read a comment block
+        private static void ReadCommentBlock(char c, Parser parser)
+        {
+            //Look for the end of the comment block
+            if (c == '/' && parser.currentString.ToString().Last() == '*')
+            {
+                parser.stateStack.Pop();
+            }
+
+            parser.currentString.Append(c);
+        }
+
+        //Get the next char from the filestream
+        private static char NextChar(Parser parser)
+        {
+            //Check if there are characters in the buffer, otherwise read a new one
+            if (parser.charBuffer.Count > 0)
+            {
+                return parser.charBuffer.Dequeue();
+            }
+            else
+            {
+                return (char)parser.fileStream.Read();
+            }
+        }
+
+        //Skip the next X characters in the filestream
+        private static void SkipChars(Parser parser, int num)
+        {
+            for (int i = 0; i < num; i++)
+            {
+                NextChar(parser);
+            }
+        }
+
+        //Utility function
+        private static string PeekString(Parser parser, int length)
+        {
+            char[] buffer = new char[length];            
```

Comment:
Why the buffering? Are you explicitly attempting to support `Stream`s that don't support seek operations, or is this just overengineering?